File: Algorithms_and_data_structure/5P/Vector.py
# Ruikang Tao
# rtao6@ucsc.edu
# programming assignment 7
# this program generates a series of functions that operate vector calculations.

"""
This module provides functions to perform standard vector operations. Vectors
are represented as lists of numbers (floats or ints). Functions that take two
vector arguments may give arbitrary output if the vectors are not compatible,
i.e. of the same dimension.
"""

import random
import math

def add(u, v):
	"""
	Return the vector sum u+v.
	"""
	vec = []
	for i in range(len(u)):
		vec.append(u[i]+v[i])
	return vec
# end add()

def negate(u):
	"""
	Return the negative -u of the vector u.
	"""
	vec = []
	for i in range(len(u)):
		vec.append(-u[i])
	return vec
# end negate()   

def sub(u, v):
	"""
	Return the vector difference u-v.
	"""
	return add(u,negate(v))
# end sub()

def scalarMult(c, u):
	"""
	Return the scalar product cu of the number c by the vector u.
	"""
	vec = []
	for i in range(len(u)):
		vec.append(c*u[i])
	return vec
# end scalarMult()

def zip(u, v):
	"""
	Return the component-wise product of u with v.
	"""
	vec = []
	for i in range(len(u)):
		vec.append(u[i]*v[i])
	return vec
# end zip()

def dot(u, v):
	"""
	Return the dot product of u with v.
	"""
	return sum(zip(u,v))
# end dot()

def length(u):
	"""
	Return the (geometric) length of the vector u.
	"""
	total = 0
	for i in range(len(u)):
		total += (u[i])**2
	#newSum = total**0.5
	return total**(0.5)
# end length()
   
def unit(v):
	"""
	Return a unit (geometric length 1) vector in the direction of v.
	"""
	vec = []
	for i in range(len(v)):
		vec.append(v[i]/length(v))
	return vec

# end unit()

def angle(u, v):
   """
   Return the angle (in degrees) between vectors u and v.
   """
   return math.degrees(math.acos(dot(unit(u),unit(v))))
# end angle()

def randVector(n, a, b):
	"""
	Return a vector of dimension n whose components are random floats
 	in the range [a, b).
 	"""
	vec = []
	for i in range(n):
		vec.append(random.uniform(a,b))
	return vec
# end randomVector()

File: Algorithms_and_data_structure/12A/Assignment 5/Vector.java
class Vector {
    public int sum;
    public int diff;
    public double mag;
    public double angle;
    public double x;
    public double y;

    public Vector(){
        int x = 0;
        int y = 0;
    }

    //set parameters
    public Vector(double i, double j){
        x=i;
        y=j;
    }

    //add brackets to the vector expression
    public String toString(){
        return("<" + this.x + "," + this.y +">");
    }

    //function calculating the sum of the vectors
    public Vector sum(Vector a){
        Vector s = new Vector(this.x + a.x, this.y + a.y);
        return s;
    }

    //fuction calculating the difference of the vectors
    public Vector diff(Vector a){
        Vector s = new Vector(this.x - a.x, this.y - a.y);
        return s;
    }

    //function calculating the magnitude of the vector
    public double mag(){
        double mag = Math.sqrt(this.x * this.x + this.y * this.y);
        return mag;

    }

    //function calculating the scalar product
    public Vector scalar(Vector a, int b){
        Vector s = new Vector(this.x * b, this.y * b);
        return s;
    }

    //function calculating the dot product
    public double dot(Vector a){
        double d = this.x * a.x + this.y * a.y;
        return d;
    }

    //function calculating the angle between two vectors
    public double angle(Vector a){
        double angle = Math.toDegrees(Math.acos( this.dot(a)/(this.mag()*a.mag()) ));
        return angle;
    }
}